### PR TITLE
Update star difficulty bindables on mod settings change

### LINKED ISSUE #9669: Improved dynamic difficulty change when mods are activated.
**Describe the new feature:** 
I think that in the future it is necessary to work out the change in complexity in the following scenarios:
1. Double time/Half time with custom speed. Now the difficulty is calculated only for the default value.
2. Difficulty adjust (circle size).
3. Wind Up/Down

### PR DESCRIPTION
**Pull request by frenzibyte**
Closes #9669 

Since there have been quite a lot of reports on this, I've decided to looking into it and it looks to be somewhat straightforward. I've separated the steps I went through to be able to support this each per commit.

I've had to apply a 100ms debounce for the beatmap difficulty cache to not be stuffed with update requests when a user starts dragging a mod setting, in a somewhat similar fashion to `AdvancedStats`.
 
https://user-images.githubusercontent.com/22781491/129656243-c337f261-8404-4123-987c-e9c32971883f.mp4

I think further improvements to this would be to have animation/counting support in `StarRatingDisplay` to feel more active with the difficulty change, but that can come along with #14128, as the current design cannot have such animation.

### CODE REVIEW

File: osu.Game/Rulesets/Mods/Mod.cs [hunk 1]
```diff
@@ -197,6 +197,18 @@ public bool Equals(Mod other)
...
         }
 
+        public override int GetHashCode()
+        {
+            var hashCode = new HashCode();
+
+            hashCode.Add(GetType());
+
+            foreach (var (_, prop) in this.GetSettingsSourceProperties())
+                hashCode.Add(ModUtils.GetSettingUnderlyingValue(prop.GetValue(this)));
```
Comment (peppy) [inline osu.Game/Rulesets/Mods/Mod.cs:207]: I'm not sure we want to be doing this in `GetHashCode`. It is supposed to be a fast operation and this seems like it could have some overhead.

File: osu.Game/Screens/Select/BeatmapInfoWedge.cs [hunk 2]
```diff
@@ -304,20 +293,22 @@ private void load(LocalisationManager localisation)
                 titleBinding.BindValueChanged(_ => setMetadata(metadata.Source));
                 artistBinding.BindValueChanged(_ => setMetadata(metadata.Source), true);
 
+                starDifficulty = difficultyCache.GetBindableDifficulty(beatmapInfo, (cancellationSource = new CancellationTokenSource()).Token);
+                starDifficulty.BindValueChanged(s =>
+                {
+                    difficultyColourBar.Current.Value = s.NewValue ?? default;
```
Comment (peppy) [inline osu.Game/Screens/Select/BeatmapInfoWedge.cs:299]: Why are these done here rather than direct binds?

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Select/BeatmapInfoWedge.cs (hunk 2 above)
Comment (frenzibyte) [reply osu.Game/Screens/Select/BeatmapInfoWedge.cs:299]: `StarRatingDisplay` and `DifficultyColourBar` both accept type `StarDifficulty`, while the returned bindable is of type `StarDifficulty?`.

File: osu.Game/Rulesets/Mods/Mod.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/Mod.cs:207]: Good point, I think we can have a static cache for these as they return the properties of the type, nothing to do with the instance itself.

File: osu.Game/Screens/Select/BeatmapInfoWedge.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Screens/Select/BeatmapInfoWedge.cs:299]: Yet they handle weird defaults locally? Wouldn't it make more sense to either make `default` used everywhere or have them support nullability? Not sure.

File: osu.Game/Rulesets/Mods/Mod.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Rulesets/Mods/Mod.cs:207]: Why are you implementing in `GetHashCode` in the first place though?

Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/Mod.cs:207]: Because it is used in `DifficultyCacheLookup.GetHashCode()`, for equality comparisons during value retrieval I believe.

File: osu.Game/Screens/Select/BeatmapInfoWedge.cs (hunk 2 above)
Comment (frenzibyte) [reply osu.Game/Screens/Select/BeatmapInfoWedge.cs:299]: Not sure either, `StarRatingDisplay` supporting nullability sounds wrong unless you're fine with making it display 0.00 or hide itself, I would expect a null value on a component to display something indicating its value.

Comment (peppy) [reply osu.Game/Screens/Select/BeatmapInfoWedge.cs:299]: Well as it stands, you're just using `default`. Doesn't that give the same behaviour?

Comment (frenzibyte) [reply osu.Game/Screens/Select/BeatmapInfoWedge.cs:299]: The `default` there won't actually be hit as the value change event doesn't invoke once immediately, I should've added a `Debug.Assert` instead.

File: osu.Game/Rulesets/Mods/Mod.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Rulesets/Mods/Mod.cs:207]: Keep in mind `GetHashCode` is a best-effort return, not the ground truth. May want to read up on why it exists. May not be required in this case.

File: osu.Game/Screens/Select/BeatmapInfoWedge.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Screens/Select/BeatmapInfoWedge.cs:299]: So the display elements have a hidden behaviour for when they don't have an initial value... which can only be reached once ever? That seems really weird.

Comment (frenzibyte) [reply osu.Game/Screens/Select/BeatmapInfoWedge.cs:299]: Only for the difficulty colour bar which is currently visible, can hide that until a star difficulty is calculated then.

File: osu.Game/Rulesets/Mods/Mod.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/Mod.cs:207]: `MemoryCachingComponent`s check for existence of computed values by doing `cache.TryGetValue(lookup, out var value)`, and the implementation of that method seems to require hash code to function properly:
CleanShot 2021-08-17 at 18 19 16

(doesn't consider equality even when `Equals` would return true)

I'll go with caching the properties just so that `GetHashCode()` doesn't get hindered by the overhead.

File: osu.Game/Screens/Select/BeatmapInfoWedge.cs (hunk 2 above)
Comment (frenzibyte) [reply osu.Game/Screens/Select/BeatmapInfoWedge.cs:299]: While I was going through making all components accept `StarDifficulty?` instead, it turns out the following will also be required to bind, unfortunately:

CleanShot 2021-08-17 at 18 47 31

I'll change to an assert and leave the rest for another PR, as otherwise things will get out of scope given that `StarRatingDisplay` only accepts `StarDifficulty` itself.

File: osu.Game/Rulesets/Mods/Mod.cs (hunk 1 above)
Comment (bdach) [reply osu.Game/Rulesets/Mods/Mod.cs:207]: Caching the properties doesn't eliminate all overhead. There's still the cost of fetching the values via reflection.

Looking at this from a different angle, I'm not really sure whether difficulty values for mods with settings *should* be cached at all. The reason why I'm doubting that is that I'm a bit skeptical as to what the hit rate of such a cache would be, especially for things like difficulty adjust where you have some hundreds of admissible *values* for each property, let alone *combinations* of values.

In other words, what I'm trying to put under consideration here is *not* caching difficulty values if they come from a mod with non-standard setting values. That would sidestep the issue entirely.

Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/Mod.cs:207]: It's not the value that's cached though? It's the property info itself, but yeah there's the overhead of fetching as well unfortunately.

Tempted to just leave this off but I'll need to benchmark because reflection is full of mysteries.

Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/Mod.cs:207]: Yeah seems there's still an overhead from the fetching alone:

- hash `GetType()` only (no properties fetching):
    |      Method | Times |        Mean |      Error |     StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
    |------------ |------ |------------:|-----------:|-----------:|------:|------:|------:|----------:|
    | ModHashCode |     1 |    19.62 ns |   0.128 ns |   0.120 ns |     - |     - |     - |         - |
    | ModHashCode |    10 |   181.62 ns |   2.087 ns |   1.629 ns |     - |     - |     - |         - |
    | ModHashCode |   100 | 2,982.04 ns | 126.045 ns | 369.668 ns |     - |     - |     - |         - |

- hash `GetType()` and fetched properties without caching:
    |      Method | Times |        Mean |     Error |    StdDev |      Median |    Gen 0 | Gen 1 | Gen 2 | Allocated |
    |------------ |------ |------------:|----------:|----------:|------------:|---------:|------:|------:|----------:|
    | ModHashCode |     1 |    55.11 us |  3.031 us |  8.842 us |    52.10 us |   2.5635 |     - |     - |      8 KB |
    | ModHashCode |    10 |   349.67 us | 29.659 us | 87.451 us |   295.46 us |  26.3672 |     - |     - |     82 KB |
    | ModHashCode |   100 | 2,729.53 us | 11.389 us |  8.891 us | 2,730.06 us | 265.6250 |     - |     - |    818 KB |

- hash `GetType()` and fetched properties with caching:
    |      Method | Times |       Mean |      Error |     StdDev |   Gen 0 | Gen 1 | Gen 2 | Allocated |
    |------------ |------ |-----------:|-----------:|-----------:|--------:|------:|------:|----------:|
    | ModHashCode |     1 |   6.074 us |  0.2061 us |  0.5946 us |  0.1526 |     - |     - |     520 B |
    | ModHashCode |    10 |  42.969 us |  0.8216 us |  2.3173 us |  1.6479 |     - |     - |   5,200 B |
    | ModHashCode |   100 | 500.425 us | 16.0374 us | 45.4956 us | 16.1133 |     - |     - |  52,000 B |

I'll look at improving this, by once-off retrieving all bindable settings rather than fetching them every equality comparsion call.

Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/Mod.cs]: After caching all bindables from first retrieval, the results appear to be the following (`GetHashCode` with no properties added and `GetHashCode` with properties added, side to side):

|                  Method | Times |         Mean |      Error |       StdDev |       Median |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------ |------ |-------------:|-----------:|-------------:|-------------:|-------:|------:|------:|----------:|
| ModHashCodeNoProperties |     1 |     29.05 ns |   0.757 ns |     2.232 ns |     29.61 ns |      - |     - |     - |         - |
|             ModHashCode |     1 |    113.39 ns |   2.233 ns |     1.979 ns |    113.21 ns | 0.0203 |     - |     - |      64 B |
| ModHashCodeNoProperties |    10 |    359.92 ns |  25.999 ns |    75.014 ns |    339.14 ns |      - |     - |     - |         - |
|             ModHashCode |    10 |  2,262.31 ns | 159.937 ns |   464.006 ns |  2,101.73 ns | 0.2022 |     - |     - |     640 B |
| ModHashCodeNoProperties |   100 |  2,932.02 ns |  70.621 ns |   201.486 ns |  2,864.35 ns |      - |     - |     - |         - |
|             ModHashCode |   100 | 17,209.31 ns | 946.256 ns | 2,730.164 ns | 16,491.80 ns | 2.0294 |     - |     - |   6,400 B |

I'm still not sure what's causing the allocation when enumerating the settings and adding them to the hashcode.

**Comment by frenzibyte** [comment]
I've added animation support to the star rating display for the beatmap info wedge to feel more active when a user changes a difficulty setting:

https://user-images.githubusercontent.com/22781491/130006990-6e57e1ff-1a6d-4f67-b131-184188da773a.mp4

The BPM changing while the star rating display is still held until after 100ms feels a bit weird, makes me think about using `TransferValueOnCommit` on all mod settings sliders and call it a day.

**Comment by peppy** [comment]
Spacing looks to have broken in the wedge:

20210819 142800 (dotnet)


Should probably be specified as part of the `FillFlowContainer`, not an individual component (https://github.com/ppy/osu/blob/25e6317e7f1a51dc6cd15edbafaccb10d499a4d1/osu.Game/Screens/Select/BeatmapInfoWedge.cs#L237)

**Comment by frenzibyte** [comment]
Resolved, was an oversight from my end.

**Review by peppy — APPROVED** [review]

**Comment by peppy** [comment]
I think this looks okay. @smoogipoo can you make sure you're okay with the hash/equality logic?

**Comment by smoogipoo** [comment]
I looked at it before and it seemed reasonable and simple, but didn't comment because I'm not sure whether the discussion above was overcomplicating things too early.
